[FLINK-26762][network] Add the overdraft buffer in ResultPartition

1996fanrui opened the pull request:
## What is the purpose of the change

Add the overdraft buffer in ResultPartition to reduce unaligned checkpoint being blocked

## Brief change log

- Add overdraft buffer in LocalBufferPool
- Add configuration : taskmanager.network.memory.max-overdraft-buffers-per-gate

## Verifying this change

This change added tests and can be verified as follows:

  - Added testOverdraftBuffer in LocalBufferPoolTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

akalash commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:
```diff
@@ -543,6 +543,164 @@ public void testConsistentAvailability() throws Exception {
     }
 
+    @Test
+    public void testOverdraftBuffer() throws IOException, InterruptedException {
```
Why do you keep all scenarios in one test? It looks too big and difficult to read. Can you please, split this test into several tests with the descriptive names?

1996fanrui commented:
> @1996fanrui , thanks for these changes. In general, the idea of implementation looks good but I have left several comments in PR and we still need to discuss configuration later.

Hi @akalash , thanks for your review. I have read your comments, I'll address them after our discussion.

I created [FLIP-227](https://cwiki.apache.org/confluence/display/FLINK/FLIP-227%3A+Support+overdraft+buffer) and raised a discussion on the [flink dev mailing list](https://lists.apache.org/thread/4p3xcf0gg4py61hsnydvwpns07d1nog7). Please help to review and add some information in dev mailing list if something is missing. Thank you very much.

1996fanrui commented:
Hi @akalash  @pnowojski , I updated this pr based on the FLIP-227, and I have addressed @akalash 's all comments. Please help to review in your free time, thanks a lot.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:
```diff
@@ -472,18 +518,22 @@ private void onGlobalPoolAvailable() {
-        return !availableMemorySegments.isEmpty() && unavailableSubpartitionsCount == 0;
+        return !availableMemorySegments.isEmpty()
+                && unavailableSubpartitionsCount == 0
+                && numberOfRequestedOverdraftMemorySegments == 0;
```
I deleted all checks in other places and use the `shouldBeAvailable()` instead of those checks to avoid misuse in the future.

1996fanrui commented:
@flinkbot run azure

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:
```diff
@@ -424,6 +452,24 @@ private boolean requestMemorySegmentFromGlobal() {
         return false;
     }
 
+    private MemorySegment requestOverdraftMemorySegmentFromGlobal(int targetChannel) {
```
It looks like `targetChannel` can be removed here since there is no usage of this parameter

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:
```diff
@@ -676,7 +726,9 @@ private void returnExcessMemorySegments() {
-        return numberOfRequestedMemorySegments > currentPoolSize;
+        return numberOfRequestedOverdraftMemorySegments > 0
+                || numberOfRequestedMemorySegments + numberOfRequestedOverdraftMemorySegments
+                        > currentPoolSize;
```
I think you overcomplicated the condition I believe this is an easier equivalent:
```
private boolean hasExcessBuffers() {
        return numberOfRequestedOverdraftMemorySegments > 0
                || numberOfRequestedMemorySegments > currentPoolSize;
    }
```

akalash reviewed (commented):
@1996fanrui ,  Thanks for these changes. I have left a couple of minor comments but in general, it looks good. I want to take some time to think about how good the current tests are and then I will be ready to approve this PR.

1996fanrui commented:
> @1996fanrui , Thanks for these changes. I have left a couple of minor comments but in general, it looks good. I want to take some time to think about how good the current tests are and then I will be ready to approve this PR.

@akalash  , thanks for your review. I have updated.

akalash commented on docs/layouts/shortcodes/generated/all_taskmanager_network_section.html:
```diff
@@ -80,6 +80,12 @@
+            <td style="word-wrap: break-word;">5</td>
+            <td>Integer</td>
+            <td>Number of max overdraft network buffers to use for each ResultPartition. The overdraft buffers will be used when the ResultPartition cannot apply to the normal buffers, e.g: the all buffers of ResultPartition be applied or the number of buffers for a single channel has reached taskmanager.network.memory.max-buffers-per-channel. When the ResultPartition is unavailable, the ResultPa ...
+        </tr>
```
Just for the check. Did you generate this documentation automatically(`mvn -Pgenerate-config-docs ...`)?

1996fanrui replied:
Yes, CI will fail if the code and doc don't match.

akalash commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:
```diff
@@ -549,6 +551,157 @@ public void testConsistentAvailability() throws Exception {
+            assertRequestedOverdraftBufferAndIsAvailable(bufferPool, i + 1, false);
+        }
+
+        BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(1);
```
I think that it makes sense to check that `requestBufferBuilder` for `0` channel returns NULL value as well

akalash commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:
```diff
@@ -549,6 +551,157 @@ public void testConsistentAvailability() throws Exception {
+        bufferPool.lazyDestroy();
+    }
+
+    private void testUseAllBuffersForMultiChannel(
```
I can be mistaken but this method looks the same as above (`testUseAllBuffersForSingleChannel`). I think you can merge them into one but where you want to use the different channels you can use (`i % numberOfChannel`)

akalash commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:
```diff
@@ -549,6 +551,157 @@ public void testConsistentAvailability() throws Exception {
+        AutoCloseableRegistry channel0CloseableRegistry = new AutoCloseableRegistry();
+        for (int i = 0; i < poolSize / 2; i++) {
+            assertRequestedOverdraftBufferAndIsAvailable(bufferPool, 0, true);
+            BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(0);
```
For example here:
```
BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(i % numberOfChannels);
```

akalash reviewed (commented):
Now it LGTM. I have left a couple of minor comments. You can fix it along with rebasing into master and after that it will be ready to merge.

1996fanrui commented:
Hi @akalash , thanks for your suggestion. Currently, the unit test is simpler than before.

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

akalash reviewed (approved):
LGTM. @pnowojski can you help with the merge, please?

pnowojski commented on flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:
```diff
@@ -273,6 +273,26 @@ public class NettyShuffleEnvironmentOptions {
+                                            + " the problem that multiple output buffers are required to process a single data, causing the Task to block in the requestMemory,"
+                                            + " such as: flatMap operator, records spanning multiple buffers or a single timer generates a large amount of data."
+                                            + " It doesn't need to wait for ResultPartition is available to start Unaligned Checkpoint directly.",
+                                    NETWORK_MAX_BUFFERS_PER_CHANNEL.key()));
```
I agree this is a bit too technical. What about:
> Number of max overdraft network buffers to use for each subtask
>
> The overdraft buffers will be used when the subtask cannot apply for the normal buffers due to back pressure,
> while subtask is performing an action that can not be interrupted in the middle, like serializing a large record,
> flatMap operator producing multiple records for one single input record or processing time timer producing large output.
> In situations like that system will allow subtask to request overdraft buffers, so that the subtask can finish such
> uninterruptible action, without blocking unaligned checkpoints for long period of time. Overdraft buffers
> are provided on best effort basis only if the system has some unused buffers available. Subtask that has used
> overdraft buffers won't be allowed to process any more records until the overdraft buffers are returned to the pool.

?

pnowojski reviewed (commented):
Thanks @akalash  and @1996fanrui . I have one minor comment and one suggestion.

pnowojski commented on docs/content.zh/docs/ops/metrics.md:
```diff
@@ -840,7 +840,7 @@ Deprecated: use [Default shuffle service metrics](#default-shuffle-service)
     <tr>
       <td>outPoolUsage</td>
-      <td>An estimate of the output buffers usage.</td>
+      <td>An estimate of the output buffers usage, the pool usage was > 100% if overdraft buffers are being used.</td>
```
nit:
```
An estimate of the output buffers usage. The pool usage can be > 100% if overdraft buffers are being used.
```

pnowojski commented on docs/content.zh/docs/ops/metrics.md:
```diff
@@ -840,7 +840,7 @@ Deprecated: use [Default shuffle service metrics](#default-shuffle-service)
+      <td>An estimate of the output buffers usage, the pool usage was > 100% if overdraft buffers are being used.</td>
       <td>Gauge</td>
     </tr>
     <tr>
```
I think in both English and Chinese versions there is also a second occurrence of the same metric for the `Default shuffle service metrics` that you also should modify. Here you have only modified the docs for the deprecated component that is preserved for the backward compatibility reasons.

1996fanrui replied:
Thanks for your reminder, updated.

pnowojski reviewed (commented):
Two more minor comments regarding the docs. Apart of that LGTM

pnowojski reviewed (approved):
LGTM :) I will merge it most-likely tomorrow with green azure.

1996fanrui commented:
@flinkbot run azure
